src/bootstrap/config.rs

```diff
@@ -910,10 +913,12 @@ impl Config {
             config.out = crate::util::absolute(&config.out);
         }
 
-        config.initial_rustc = build
-            .rustc
-            .map(PathBuf::from)
-            .unwrap_or_else(|| config.out.join(config.build.triple).join("stage0/bin/rustc"));
+        config.initial_rustc = build.rustc.map(PathBuf::from).unwrap_or_else(|| {
+            let build = crate::Build::new(config.clone());
+            let builder = Builder::new(&build);
+            download_beta_toolchain(&builder);
```

Comment:
This feels pretty suspicious. I don't think we should need to construct a full Build/Builder to download these binaries, what are they depending on that causes this? Can we move those functions onto Config?

Reply:
There's a *lot* of functions defined on Builder but not Config; no objection to moving them, but it's quite a large diff. We need to move:
- fix_bin_or_dylib
- download_component
- `verbose` (I'd keep a helper on Builder that calls Config::verbose for this)
- download_http_with_retries
- unpack
- verify
- tempdir
- `try_run` (I'd keep a helper)
- `check_run` (helper)
- `create` (helper)
- `remove` (helper)

and we might need some more after that, I don't know everything those functions call internally.

Hmm, I guess we need all that anyway to avoid pulling `Builder` into the shim; ok, seems worthwhile then.